yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts

```diff
@@ -300,4 +342,181 @@ export class AztecKVTxPool implements TxPool {
+      numNewTxsEvicted += newTxHashes.filter(txHash => txsToEvict.includes(txHash)).length;
+
+      pendingTxsByPriority = pendingTxsByPriority.slice(numTxsToEvict);
+    }
+
+    let pendingTxsSize = (await this.#pendingTxSize.getAsync()) ?? 0;
+    if (pendingTxsSize > this.#maxTxPoolSize) {
+      for (const txHash of pendingTxsByPriority) {
+        txsToEvict.push(txHash);
+
+        const txSize = (await this.#pendingTxHashToSize.getAsync(txHash.toString())) ?? 0;
+        pendingTxsSize -= txSize;
+        if (pendingTxsSize <= this.#maxTxPoolSize) {
+          break;
+        }
+      }
+      numNewTxsEvicted += newTxHashes.filter(txHash => txsToEvict.includes(txHash)).length;
+    }
+
+    if (txsToEvict.length > 0) {
+      await this.deleteTxs(txsToEvict, true);
+    }
+    return {
+      numLowPriorityTxsEvicted: txsToEvict.length,
+      numNewTxsEvicted,
+    };
+  }
+
+  /**
+   * Evicts invalid pending txs from the pool after a txs from a block are mined.
+   * Eviction criteria includes:
+   *   - txs with nullifiers that are already included in the mined block
+   *   - txs with an insufficient fee payer balance
+   *   - txs with a max block number lower than the mined block
+   *
+   * @param minedTxHashes - The tx hashes of the txs mined in the block.
+   * @param blockNumber - The block number of the mined block.
+   * @returns The total number of txs evicted from the pool.
+   *
+   * TODO: add logging for evicted txs?
+   */
+  private async evictInvalidTxsAfterMining(minedTxHashes: TxHash[], blockNumber: number): Promise<number> {
+    if (minedTxHashes.length === 0) {
+      return 0;
+    }
+
+    const pendingTxHashes = await this.getPendingTxHashes();
+    const minedTxs = await Promise.all(minedTxHashes.map(async hash => await this.getTxByHash(hash)));
+    const minedNullifiers = new Set<Fr>();
+    for (const tx of minedTxs) {
+      if (tx) {
+        const nullifiers = tx.data.getNonEmptyNullifiers();
+        nullifiers.forEach(nullifier => minedNullifiers.add(nullifier));
+      }
+    }
+    const pendingTxs = await Promise.all(
+      pendingTxHashes.map(async hash => ({
+        hash,
+        tx: await this.getTxByHash(hash),
+      })),
+    );
+
+    const txsToEvict: TxHash[] = [];
+    for (const { hash, tx } of pendingTxs) {
+      if (!tx) {
+        continue;
+      }
+
+      // Evict pending txs that share nullifiers with mined txs
+      const txNullifiers = tx.data.getNonEmptyNullifiers();
+      if (txNullifiers.some(nullifier => minedNullifiers.has(nullifier))) {
+        txsToEvict.push(hash);
+        continue;
+      }
+
+      // Evict pending txs with an insufficient fee payer balance
+      const db = this.#worldStateSynchronizer.getCommitted();
+      const gasTxValidator = new GasTxValidator(
+        new DatabasePublicStateSource(db),
+        ProtocolContractAddress.FeeJuice,
+        GasFees.empty(),
+      );
```

Comment:
Let's initialize this out of the tx loop, IIRC getting a world state view is not cheap

Reply:
Also, we need to make sure world state is synchronized to the block we're processing before calling `getCommitted`. Both p2p and world state sync from the archiver at their own pace. If `markAsMined` gets called and reaches this point _before_ world state had time to sync, we'll be seeing the old balances.